programs/move_loader_api/src/lib.rs

```diff
@@ -215,186 +248,279 @@ mod tests {
+        assert_eq!(0, AccountResource::read_sequence_number(&payee_resource));
+    }
+
+    // Helpers
+
+    fn mint_coins(amount: u64) -> Result<Vec<LibraAccount>, InstructionError> {
+        let code = "
+            import 0x0.LibraAccount;
+            import 0x0.LibraCoin;
+            main(payee: address, amount: u64) {
+                LibraAccount.mint_to_address(move(payee), move(amount));
+                return;
+            }
+        ";
+        let mut genesis = LibraAccount::create_genesis();
+        let mut program = LibraAccount::create_program(&genesis.address, code);
+        let mut payee = LibraAccount::create_unallocated();
 
         let mut keyed_accounts = vec![
-            KeyedAccount::new(&move_program_pubkey, false, &mut move_program_account),
-            KeyedAccount::new(&genesis_pubkey, false, &mut genesis_account),
-            KeyedAccount::new(&receiver_pubkey, false, &mut receiver_account),
+            KeyedAccount::new(&program.key, false, &mut program.account),
+            KeyedAccount::new(&genesis.key, false, &mut genesis.account),
+            KeyedAccount::new(&payee.key, false, &mut payee.account),
         ];
+        let invoke_info = InvokeInfo {
+            sender_address: genesis.address.clone(),
+            args: vec![
+                TransactionArgument::Address(pubkey_to_address(&payee.key)),
+                TransactionArgument::U64(amount),
+            ],
+        };
+
+        call_process_instruction(&mut keyed_accounts, invoke_info);
 
-        let receiver_addr = AccountAddress::new(to_array_32(receiver_pubkey.as_ref()).clone());
-        let mint_amount = 42;
-        let mut args: Vec<TransactionArgument> = Vec::new();
-        args.push(TransactionArgument::Address(receiver_addr));
-        args.push(TransactionArgument::U64(mint_amount));
-        let data = bincode::serialize(&args).unwrap();
+        Ok(vec![
+            LibraAccount::new(program.key, program.account),
+            LibraAccount::new(genesis.key, genesis.account),
+            LibraAccount::new(payee.key, payee.account),
+        ])
+    }
+
+    fn call_process_instruction(keyed_accounts: &mut [KeyedAccount], invoke_info: InvokeInfo) {
+        let program_id = Pubkey::new(&MOVE_LOADER_PROGRAM_ID);
 
+        let data = bincode::serialize(&invoke_info).unwrap();
         let ix = LoaderInstruction::InvokeMain { data };
         let ix_data = bincode::serialize(&ix).unwrap();
 
-        process_instruction(&program_id, &mut keyed_accounts, &ix_data).unwrap();
+        process_instruction(&program_id, keyed_accounts, &ix_data).unwrap();
+    }
 
-        let mut data_store = DataStore::default();
-        match bincode::deserialize(&keyed_accounts[2].account.data).unwrap() {
-            LibraAccountState::Unallocated => panic!("Invalid account state"),
-            LibraAccountState::WriteSet(write_set) => data_store.apply_write_set(&write_set),
+    struct LibraAccount {
```

Comment:
Hard to get my head around this. Weird to see something called `Account` that includes a key to itself. Seems like this code could be a set of functions that each return a `KeyedAccount`.

Reply:
KeyedAccounts are ok if you are passed one but not that great to construct due to the references it contains